pkg/services/statefetcher/statefetcher.go

```diff
@@ -0,0 +1,385 @@
+package statefetcher
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"strconv"
+	"sync"
+	"sync/atomic"
+
+	"github.com/nspcc-dev/neo-go/pkg/config"
+	"github.com/nspcc-dev/neo-go/pkg/core/storage"
+	gio "github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/services/helpers/neofs"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neofs-sdk-go/client"
+	"github.com/nspcc-dev/neofs-sdk-go/container"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	"github.com/nspcc-dev/neofs-sdk-go/object"
+	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
+	"go.uber.org/zap"
+)
+
+// Ledger is the interface for statefetcher.
+type Ledger interface {
+	GetConfig() config.Blockchain
+	HeaderHeight() uint32
+	AddContractStorageData(kv []storage.KeyValue, syncHeight uint32, expectedRoot util.Uint256) error
+	GetLastStoredKey() []byte
+}
+
+// Service fetches contract storage state from NeoFS.
+type Service struct {
+	neofs.BasicService
+	containerMagic int
+
+	isActive   atomic.Bool
+	isShutdown atomic.Bool
+	lock       sync.RWMutex
```

Comment:
Move this lock prior to `lastStateObjectIndex` and `lastStateOid`, move them to a separate group.